pkg/engine/query.go

```diff
@@ -0,0 +1,196 @@
+	Diag       diag.Sink    // the sink to use for diag'ing.
+	StatusDiag diag.Sink    // the sink to use for diag'ing status messages.
+	host       plugin.Host  // the plugin host to use for this query.
+	pwd, main  string
+	plugctx    *plugin.Context
+}
+
+func Query(ctx *Context, u UpdateInfo, opts UpdateOptions) result.Result {
+	contract.Require(u != nil, "update")
+	contract.Require(ctx != nil, "ctx")
+
+	defer func() { ctx.Events <- cancelEvent() }()
+
+	tracingSpan := func(opName string, parentSpan opentracing.SpanContext) opentracing.Span {
+		// Create a root span for the operation
+		opts := []opentracing.StartSpanOption{}
+		if opName != "" {
+			opts = append(opts, opentracing.Tag{Key: "operation", Value: opName})
+		}
+		if parentSpan != nil {
+			opts = append(opts, opentracing.ChildOf(parentSpan))
+		}
+		return opentracing.StartSpan("pulumi-query", opts...)
+	}("query", ctx.ParentSpan)
+	defer tracingSpan.Finish()
+
+	emitter, err := makeEventEmitter(ctx.Events, u)
+	if err != nil {
+		return result.FromError(err)
+	}
+
+	// First, load the package metadata and the deployment target in preparation for executing the package's program
+	// and creating resources.  This includes fetching its pwd and main overrides.
+	diag := newEventSink(emitter, false)
+	statusDiag := newEventSink(emitter, true)
+
+	proj, target := u.GetProject(), u.GetTarget()
+	contract.Assert(proj != nil)
+	contract.Assert(target != nil)
+
+	pwd, main, plugctx, err := ProjectInfoContext(&Projinfo{Proj: proj, Root: u.GetRoot()}, opts.host,
+		target, diag, statusDiag, tracingSpan)
+	if err != nil {
+		return result.FromError(err)
+	}
+
+	return query(ctx, u, QueryOptions{
+		Events:     emitter,
+		Diag:       diag,
+		StatusDiag: statusDiag,
+		host:       opts.host,
+		pwd:        pwd,
+		main:       main,
+		plugctx:    plugctx,
+	})
+}
+
+func newQuerySource(cancel context.Context, client deploy.BackendClient, u UpdateInfo,
+	opts QueryOptions) (deploy.QuerySource, error) {
+
+	// Before launching the source, ensure that we have all of the plugins that we need in order to proceed.
+	//
+	// There are two places that we need to look for plugins:
+	//   1. The language host, which reports to us the set of plugins that the program that's about to execute
+	//      needs in order to create new resources. This is purely advisory by the language host and not all
+	//      languages implement this (notably Python).
+	//   2. The snapshot. The snapshot contains plugins in two locations: first, in the manifest, all plugins
+	//      that were loaded are recorded. Second, all first class providers record the version of the plugin
+	//      to which they are bound.
+	//
+	// In order to get a complete view of the set of plugins that we need for an query, we must consult both
+	// sources and merge their results into a list of plugins.
+	languagePlugins, err := gatherPluginsFromProgram(opts.plugctx, plugin.ProgInfo{
```

Comment:
We're copy/pasting a *lot* of code here.  Can we share some/all of this better? cc @swgillespie for thoughts on structure.

Reply:
We are. We chose to do it this way rather than to impact the update path. Open to other suggestions, but I actually think of this as a feature, not a bug.

Reply:
I'll defer to @swgillespie and @ellismg - but it feels like we should extract some of this stuff out and reuse it.  All of the plugin loading here is something that should be usable in both.  It doesn't feel helpful to me to have copies of this.

Reply:
oh you mean just the plugin stuff (vs the entire PR)? We can probably consolidate that, yes.